node/header_exchange.go

```diff
@@ -0,0 +1,241 @@
+		return err
+	}
+
+	// Check if the headerstore is not initialized and try initializing
+	if hExService.headerStore.Height() > 0 {
+		if err := hExService.syncer.Start(hExService.ctx); err != nil {
+			return fmt.Errorf("error while starting the syncer: %w", err)
+		}
+		hExService.syncerStarted = true
+		return nil
+	}
+
+	// Look to see if trusted hash is passed, if not get the genesis header
+	var trustedHeader *types.Header
+	// Try fetching the trusted header from peers if exists
+	if len(peerIDs) > 0 {
+		if hExService.conf.TrustedHash != "" {
+			trustedHashBytes, err := hex.DecodeString(hExService.conf.TrustedHash)
+			if err != nil {
+				return fmt.Errorf("fail to parse the trusted hash for initializing the headerstore: %w", err)
+			}
+
+			if trustedHeader, err = hExService.ex.Get(hExService.ctx, header.Hash(trustedHashBytes)); err != nil {
+				return fmt.Errorf("fail to fetch the trusted header for initializing the headerstore: %w", err)
+			}
+		} else {
+			// Try fetching the genesis header if available, otherwise fallback to signed headers
+			if trustedHeader, err = hExService.ex.GetByHeight(hExService.ctx, uint64(hExService.genesis.InitialHeight)); err != nil {
+				// Full/light nodes have to wait for aggregator to publish the genesis header
+				// if the aggregator is passed as seed while starting the fullnode
+				return fmt.Errorf("failed to get genesis header: %w", err)
+			}
+		}
+	}
+	go hExService.tryInitHeaderStoreAndStartSyncer(hExService.ctx, trustedHeader)
+
+	return nil
+}
+
+// OnStop is a part of Service interface.
+func (hExService *HeaderExchangeService) Stop() error {
+	err := hExService.headerStore.Stop(hExService.ctx)
+	err = multierr.Append(err, hExService.p2pServer.Stop(hExService.ctx))
+	err = multierr.Append(err, hExService.ex.Stop(hExService.ctx))
+	err = multierr.Append(err, hExService.sub.Stop(hExService.ctx))
+	if !hExService.conf.Aggregator && hExService.syncerStarted {
+		err = multierr.Append(err, hExService.syncer.Stop(hExService.ctx))
+	}
+	return err
+}
+
+// newP2PServer constructs a new ExchangeServer using the given Network as a protocolID suffix.
+func newP2PServer(
+	host host.Host,
+	store *goheaderstore.Store[*types.Header],
+	network string,
+	opts ...goheaderp2p.Option[goheaderp2p.ServerParameters],
+) (*goheaderp2p.ExchangeServer[*types.Header], error) {
+	return goheaderp2p.NewExchangeServer[*types.Header](host, store, network, opts...)
+}
+
+func newP2PExchange(
+	host host.Host,
+	peers []peer.ID,
+	network string,
+	conngater *conngater.BasicConnectionGater,
+	opts ...goheaderp2p.Option[goheaderp2p.ClientParameters],
+) (*goheaderp2p.Exchange[*types.Header], error) {
+	return goheaderp2p.NewExchange[*types.Header](host, peers, network, conngater, opts...)
+}
+
+// InitStore is a type representing initialized header store.
+// NOTE: It is needed to ensure that Store is always initialized before Syncer is started.
+type InitStore header.Store[*types.Header]
```

Comment:
Do we really need this? I see it's a celestia-node influenced type, but I think they only introduced it because of use of `fx` / dependency injection.